File: problems/prob7 - 10001st prime/sln.go
package main

import (
	"fmt"
	"math"
)

func main() {
	seive := sieve_of_eratosthenes(500000)
	fmt.Println(seive[10001])
}

func sieve_of_eratosthenes(n int) []int {
	sieve := make_range(n)
	for i := 2; i < int(math.Sqrt(float64(n))); i++ {
		criteria := func(y int) bool {
			return y == i || y % i != 0
		}
		sieve = filter(sieve, criteria)
	}
	return sieve;
}

func make_range(size int) []int {
	arr := make([]int, size)
	for i := range arr {
		arr[i] = i
	}
	return arr
}

func filter(arr []int, criteria func(int) bool) []int {
	ret := make([]int, 0)
	for _, i := range arr {
		if criteria(i) {
			ret = append(ret, i)
		}
	}
	return ret
}

File: problems/prob7 - 10001st prime/sln.py
from typing import List
import math


def sieve_of_eratosthenes(n: int) -> List[int]:
    sieve = range(n)
    for i in range(2, math.ceil(math.sqrt(n))):
        criteria = lambda y: y == i or y % i != 0
        sieve = list(filter(criteria, sieve))
    
    return sieve


if __name__ == "__main__":
    seive = sieve_of_eratosthenes(500000)
    print(seive[10001])
